### PR TITLE
[New Feature] Add Focal loss for imbalanced binary classification tasks.

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/model/loss_func.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
 import torch as th
 from torch import nn
 import torch.nn.functional as F
+import torchvision
```
Comment (thvasilo) [inline python/graphstorm/model/loss_func.py:21]: Is torchvision part of our dependency list?

Comment (classicsong) [reply python/graphstorm/model/loss_func.py:21]: I need to check.

Comment (classicsong) [reply python/graphstorm/model/loss_func.py:21]: torchvision is not in our dependency list.
I have updated the implementation.

Comment (thvasilo) [reply python/graphstorm/model/loss_func.py:21]: OK can we remove this line then?

File: python/graphstorm/config/argument.py [hunk 2]
```diff
@@ -3091,8 +3112,16 @@ def _add_link_prediction_args(parser):
...
+            help="Common hyperparameter symbol gamma."
+    )
+    group.add_argument(
+            "--alpha",
+            type=float,
+            default=argparse.SUPPRESS,
+            help="Common hyperparameter symbol alpha."
     )
+    group.add_argument("--class-loss-func", type=str, default=argparse.SUPPRESS,
+            help="Classification loss function.")
```
Comment (thvasilo) [inline python/graphstorm/config/argument.py:3124]: Can we add choices here like 
```suggestion
    group.add_argument("--class-loss-func", type=str, default=argparse.SUPPRESS,
            help="Classification loss function.", choices=BUILTIN_CLASS_LOSS_FUNCTION)
```

File: docs/source/cli/model-training-inference/configuration-run.rst [hunk 3]
```diff
@@ -474,11 +474,21 @@ Link Prediction Task
...
     - Yaml: ``gamma: 10.0``
     - Argument: ``--gamma 10.0``
-    - Default value: ``12.0``
+    - Default value: None
+- **alpha**: Set the value of the hyperparameter denoted by the symbol alpha. Alpha is used in the following case: i/ focal loss for binary classification.
+
+    - Yaml: ``alpha: 10.0``
+    - Argument: ``--alpha 10.0``
+    - Default value: None
+- **class_loss_func**: Node/Edge classification loss function. Builtin loss functions include ``cross_entropy`` and ``focal``.
```
Comment (thvasilo) [inline docs/source/cli/model-training-inference/configuration-run.rst:487]: Can we add some links and explanations about what the alpha and gamma params control for each function?

File: python/graphstorm/model/edge_decoder.py [hunk 4]
```diff
@@ -448,7 +454,10 @@ def predict(self, g, h, e_h=None):
         elif self.multilabel:
             out = (th.sigmoid(out) > .5).long()
         else:  # not multilabel
-            out = out.argmax(dim=1)
+            if self.out_dim == 1:
+                out = (th.sigmoid(out) > .5).long()
+            else:
+                out = out.argmax(dim=1)
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/model/edge_decoder.py:460]: Related to focal loss.
Focal loss only works with binary classification and expects the prediction output with out-dim of 1.

File: python/graphstorm/config/argument.py (hunk 2 above)
Comment (classicsong) [reply python/graphstorm/config/argument.py:3124]: We have a check in L2525. 
The launch.py script may not be able to capture the argument error, but it can capture the assert error for sure.

File: docs/source/cli/model-training-inference/configuration-run.rst (hunk 3 above)
Comment (classicsong) [reply docs/source/cli/model-training-inference/configuration-run.rst:487]: It is better to have a page about how to tune GraphStorm models with different settings.
Let's create a new task for it.

Comment (thvasilo) [reply docs/source/cli/model-training-inference/configuration-run.rst:487]: Added this https://github.com/awslabs/graphstorm/issues/1004

File: docs/source/cli/model-training-inference/configuration-run.rst [hunk 5]
```diff
@@ -474,11 +474,21 @@ Link Prediction Task
     - Yaml: ``lp_decoder_type: dot_product``
     - Argument: ``--lp-decoder-type dot_product``
     - Default value: ``dot_product``
-- **gamma**: Gamma for ``DistMult``. The margin value in the score function.
+- **gamma**: Set the value of the hyperparameter denoted by the symbol gamma. Gamma is used in the following cases: i/ focal loss for binary classification ii/ DistMult score function for link prediction and iii/ RotatE score function for link prediction.
 
     - Yaml: ``gamma: 10.0``
     - Argument: ``--gamma 10.0``
-    - Default value: ``12.0``
+    - Default value: None
```
Comment (thvasilo) [inline docs/source/cli/model-training-inference/configuration-run.rst:481]: Can we list the defaults per loss function?

File: docs/source/cli/model-training-inference/configuration-run.rst [hunk 6]
```diff
@@ -474,11 +474,21 @@ Link Prediction Task
...
+- **alpha**: Set the value of the hyperparameter denoted by the symbol alpha. Alpha is used in the following case: i/ focal loss for binary classification.
+
+    - Yaml: ``alpha: 10.0``
+    - Argument: ``--alpha 10.0``
+    - Default value: None
```
Comment (thvasilo) [inline docs/source/cli/model-training-inference/configuration-run.rst:486]: Same here let's list defaults per loss

File: python/graphstorm/gsf.py [hunk 7]
```diff
@@ -473,9 +498,21 @@ def create_builtin_edge_decoder(g, decoder_input_dim, config, train_task):
...
+                                         config.multilabel_weights,
+                                         config.imbalance_class_weights)
+        elif config.class_loss_func == BUILTIN_CLASS_LOSS_FOCAL:
+            # set default value of alpha to 0.25 for focal loss
+            # set default value of gamma to 2. for focal loss
+            alpha = config.alpha if config.alpha is not None else 0.25
+            gamma = config.gamma if config.gamma is not None else 2.
+            loss_func = FocalLossFunc(alpha, gamma)
+        else:
+            raise RuntimeError("Unknow classification loss %s",
```
Comment (thvasilo) [inline python/graphstorm/gsf.py:513]:
```suggestion
            raise RuntimeError("Unknown classification loss %s",
```

File: python/graphstorm/gsf.py [hunk 8]
```diff
@@ -618,34 +655,40 @@ def create_builtin_lp_decoder(g, decoder_input_dim, config, train_task):
...
-                                         config.gamma)
+                                         gamma)
         else:
             decoder = LinkPredictWeightedRotatEDecoder(g.canonical_etypes,
                                                        decoder_input_dim,
-                                                       config.gamma,
+                                                       gamma,
                                                        config.lp_edge_weight_for_loss)
     else:
         raise Exception(f"Unknow link prediction decoder type {config.lp_decoder_type}")
```
Comment (thvasilo) [inline python/graphstorm/gsf.py:694]:
```suggestion
        raise Exception(f"Unknown link prediction decoder type {config.lp_decoder_type}")
```

File: docs/source/cli/model-training-inference/configuration-run.rst (hunk 6 above)
Comment (classicsong) [reply docs/source/cli/model-training-inference/configuration-run.rst:486]: I will create another PR for that.

**Review by thvasilo — APPROVED** [review]
LGTM now

File: python/graphstorm/model/loss_func.py [hunk 9]
```diff
@@ -71,6 +71,73 @@ def out_dims(self):
...
+    Parameters
+    ----------
+    alpha: float
+        Weighting factor in range (0,1) to balance
+        ositive vs negative examples or -1 for ignore. Default: ``0.25``.
+    gamma: float
+        Exponent of the modulating factor (1 - p_t) to
+        balance easy vs hard examples. Default: ``2``.
+    """
+    def __init__(self, alpha=0.25, gamma=2):
```
Comment (Oxfordblue7) [inline python/graphstorm/model/loss_func.py:90]: The default alpha and gamma are set in gsf.py, do we need to set them here as well?

File: python/graphstorm/gsf.py [hunk 10]
```diff
@@ -357,9 +361,20 @@ def create_builtin_node_decoder(g, decoder_input_dim, config, train_task):
...
                                        norm=config.decoder_norm)
-            loss_func = ClassifyLossFunc(config.multilabel,
-                                         config.multilabel_weights,
-                                         config.imbalance_class_weights)
+            if config.class_loss_func == BUILTIN_CLASS_LOSS_CROSS_ENTROPY:
+                loss_func = ClassifyLossFunc(config.multilabel,
+                                             config.multilabel_weights,
+                                             config.imbalance_class_weights)
+            elif config.class_loss_func == BUILTIN_CLASS_LOSS_FOCAL:
+                assert config.num_classes == 1, "Focal loss only works with binary classification."
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gsf.py:369]: Focal loss only requires the num_classes to be 1 as it out_dim is supposed to be 1.

File: python/graphstorm/model/loss_func.py (hunk 9 above)
Comment (classicsong) [reply python/graphstorm/model/loss_func.py:90]: I just keep the default value here.

File: python/graphstorm/gsf.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Oxfordblue7) [reply python/graphstorm/gsf.py:369]: I was referring to the assertion message. Should we specify something like num_classes must be 1, instead of saying binary classification?

[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gsf.py:369]: I c.
